### ISSUE TITLE
v5 ILiteCollection sharing between threads

### ISSUE THREAD

**Issue by martin-slater** [litedb-org/LiteDB#2084]
In v4, ILiteCollection was not listed as being thread safe, is this still the case in V5?   Can I retrieve it once and safely use the same instance from multiple threads, potentially concurrently?

cheers, Martin.

**Comment by JKamsker** [comment]
Yes. In v5 (5.0.21, and the same on the 6.x prereleases) you can get a collection once and use that instance from many threads concurrently:

- `LiteCollection<T>` only holds read-only state: the collection name, the engine, the mapper and the entity mapping.
- Every call is forwarded to the engine, which is thread-safe. Reads use snapshots, and writes are serialized per collection.
- `Include(...)` doesn't modify the shared instance. It returns a **new** collection object with the include added.

```csharp
// e.g. held in a singleton service
private readonly ILiteCollection<Order> _orders = db.GetCollection<Order>("orders");

// safe from any thread
_orders.Insert(order);
var recent = _orders.Find(x => x.Customer == "c3").ToList();
var withRefs = _orders.Include(x => x.Customer).FindById(id); // new collection, shared one untouched
```

Keep a single `LiteDatabase` per file in the process, which the collection belongs to (see #1412). Also materialize query results with `ToList()` instead of sharing an open enumerator between threads.

One caveat for 5.0.21: it has known `BsonMapper` races when several threads map a type for the first time at the same moment (#2528). The fixes are on `dev` and in the 6.x prereleases, but not in a 5.0.x NuGet release yet.

Closing this as answered — if it doesn't cover your case, comment here and we'll reopen it.
